leakcanary-android-instrumentation/src/main/java/com/squareup/leakcanary/FailTestOnLeakRunListener.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright (C) 2018 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.leakcanary;
+
+import android.app.Application;
+import android.app.Instrumentation;
+import android.content.Context;
+import android.os.Bundle;
+import android.support.test.internal.runner.listener.InstrumentationResultPrinter;
+import org.junit.runner.Description;
+import org.junit.runner.notification.Failure;
+import org.junit.runner.notification.RunListener;
+
+import static android.support.test.InstrumentationRegistry.getInstrumentation;
+import static android.support.test.internal.runner.listener.InstrumentationResultPrinter.REPORT_VALUE_RESULT_FAILURE;
+
+/**
+ * <p>A JUnit {@link RunListener} for detecting memory leaks in Android instrumentation tests. It
+ * waits for the end of a test, and if the test succeeds then it will look for leaking
+ * references, trigger a heap dump if needed and perform an analysis.
+ *
+ * <p>To use it, you need to:
+ * <ul>
+ * <li>Install a custom RefWatcher that will not trigger heapdumps while the tests run.</li>
+ * <li>Add this {@link RunListener} as an instrumentation test listener.</li>
+ * </ul>
+ *
+ * <h3>Installing the instrumentation RefWatcher</h3>
+ *
+ * <p>For this test listener to work correctly, the {@link RefWatcher} must keep track of
+ * references but not trigger any heap dump until this test listener runs, otherwise an analysis in
+ * progress might prevent this listener from performing its own analysis.
+ *
+ * <p>Create and install the {@link RefWatcher} instance using
+ * {@link InstrumentationLeakDetector#instrumentationRefWatcher(Application)} instead of
+ * {@link LeakCanary#install(Application)} or {@link LeakCanary#refWatcher(Context)}.
+ * <pre><code>
+ * public class TestApplication extends ExampleApplication {
+ *   {@literal @}Override protected void setupLeakCanary() {
+ *     InstrumentationLeakDetector.instrumentationRefWatcher(this)
+ *         .buildAndInstall();
+ *   }
+ * }
+ * </code></pre>
+ *
+ * <h3>Add FailTestOnLeakRunListener as an intrumentation test listener</h3>
+ *
+ * <p>You can add the following to the config of your build.gradle:
+ *
+ * <pre><code>testInstrumentationRunnerArgument "listener", "com.squareup.leakcanary.FailTestOnLeakRunListener"</code></pre>
+ *
+ * <p>If instead you are running UI tests manually, add a <em>listener</em> execution argument to
```

Comment:
by "manually" do you mean from the command line? what is the preferred/other way to run this?

Reply:
typical way would be something like `./gradlew leakcanary-sample:connectedCheck`, which requires having the testInstrumentationRunnerArgument in your gradle file. I'll update the comments.